### PR TITLE
[v1] Move block_hashes from KVCacheManager to Request.block_hashes (#19728)

### PR DESCRIPTION
**Pull request by orozery**
This PR move the request block hashes from the KVCacheManager to the Request object itself.
In particular, this will allow connectors to access the request block hashes.

Part of the work described in RFC #19854

### CODE REVIEW

**Review by njhill — COMMENTED** [review]
This looks reasonable to me but would be good to get approval from @heheda12345
 and @WoosukKwon.

File: vllm/v1/request.py [hunk 1]
```diff
@@ -18,6 +18,20 @@
     from vllm.lora.request import LoRARequest
 
 
+class BlockHash(NamedTuple):
```
Comment (heheda12345) [inline vllm/v1/request.py]: Prefer to still put BlockHash in `kv_cache_utils` as there are many other data structures related to block hash there.

**Review by heheda12345 — COMMENTED** [review]
My major concern is that the `block_hash` list is updated here https://github.com/vllm-project/vllm/blob/d3f05c9248d79dc900c79d090db16cc2e5d96ee3/vllm/v1/core/block_pool.py#L177  
It is quite far from `Request` class. It will make the number of element in the list magic to connectors, and hurt the modularity of the code.
Therefore, if you want to put it into `Request` class, can you move this logic from "update when block hash is needed" to "update when token ids of the request are updated"?

**Comment by orozery** [comment]
> My major concern is that the `block_hash` list is updated here
> 
> https://github.com/vllm-project/vllm/blob/d3f05c9248d79dc900c79d090db16cc2e5d96ee3/vllm/v1/core/block_pool.py#L177
> 
> 
> It is quite far from `Request` class. It will make the number of element in the list magic to connectors, and hurt the modularity of the code.
> Therefore, if you want to put it into `Request` class, can you move this logic from "update when block hash is needed" to "update when token ids of the request are updated"?

Today, `block_hashes` are computed when the scheduler calls `_update_waiting_for_remote_kv`.
Shortly after this call, calls are made to `kv_cache_manager.get_computed_blocks` and `connector.get_num_new_matched_tokens`.
The connector must get the block hashes at this point.

I'm trying to interpret your suggestion "update when token ids of the request are updated".
My best guess is that you mean `request.append_output_token_ids`. This is too late for the connector.
I'm guessing that I do not understand you.
Can you please be more concrete and reference to actual code locations?

**Comment by heheda12345** [comment]
Yes I mean `request.append_output_token_ids`.
Based on the code here, the new tokens are updated in `update_from_output` in step k, and then used by the connector in `scheduler.schedule()` of step `k+1`. Why is it too late?
https://github.com/vllm-project/vllm/blob/72d14d0eed4b29e5827519283c085a7a674f3256/vllm/v1/engine/core.py#L223-L240

**Comment by orozery** [comment]
> Yes I mean `request.append_output_token_ids`. Based on the code here, the new tokens are updated in `update_from_output` in step k, and then used by the connector in `scheduler.schedule()` of step `k+1`. Why is it too late?
> 
> https://github.com/vllm-project/vllm/blob/72d14d0eed4b29e5827519283c085a7a674f3256/vllm/v1/engine/core.py#L223-L240

@heheda12345 Consider the following scenario:
1. Engine gets a fresh new prompt with 1000 (number does not matter) tokens.
2. Scheduler pops the prompt from its `self.waiting` list.
3. Scheduler checks whether this prompt has already computed tokens in the GPU prefix cache, by calling `self.kv_cache_manager.get_computed_blocks`. Assumes no hits are found, so `num_computed_tokens = 0`.
4. Next, it queries the connector of any externally hit tokens, by calling `self.connector.get_num_new_matched_tokens(request, ...)`. At this point, the connector needs the block hashes in order to check for hits.
5. Assuming your approach of delaying the setting of `request.block_hashes`, the connector will yield `num_external_computed_tokens = 0`, even though it may have all the tokens available!
6. The request tokens will be sent to the workers, and the workers will recompute the KV-values for this 1000 tokens.
7. Only after the workers re-compute the tokens, the scheduler will call `update_from_output`, setting the block_hashes.

So to sum-up, because of the delay in setting of `request.block_hashes`, we lose the ability to utilize offloaded KV-values of tokens from external source.

**Comment by heheda12345** [comment]
But I think the hash of the prompt can be initialized when the request is created

**Comment by orozery** [comment]
> But I think the hash of the prompt can be initialized when the request is created

@heheda12345 The block hashes are computed inside `BlockPool.cache_full_blocks`.
The `Request` object is created at `EngineCore.add_request`.
Are you suggesting that we compute the hashes inside `EngineCore.add_request` ?

**Comment by heheda12345** [comment]
Yes you can compute the hash in both `add_request` and `append_output_token_ids`. Though I prefer the solution in https://github.com/vllm-project/vllm/pull/15652 for hashing prompt tokens , I'm OK with computing hash in `add_request` in this PR and move it to frontend later.

**Comment by orozery** [comment]
> Yes you can compute the hash in both `add_request` and `append_output_token_ids`.

@heheda12345 I made the changes as you suggested.

File: vllm/v1/core/sched/scheduler.py [hunk 2]
```diff
@@ -925,6 +930,27 @@ def _update_request_with_output(
...
+                if request.block_hashes:
+                    prev_block_hash_value = (
+                        request.block_hashes[-1].hash_value)
+
+                block_tokens = request.all_token_ids[
+                    start_token_idx:end_token_idx]
+                block_hash = hash_block_tokens(self.caching_hash_fn,
+                                               prev_block_hash_value,
+                                               block_tokens, extra_keys)
+                request.block_hashes.append(block_hash)
```
Comment (heheda12345) [inline vllm/v1/core/sched/scheduler.py]: is it possible to put hash update into append_output_token_ids?

File: vllm/v1/request.py [hunk 3]
```diff
@@ -82,6 +83,8 @@ def __init__(
         self.num_computed_tokens = 0
         self.cache_salt: Optional[str] = cache_salt
 
+        self.block_hashes: list[BlockHash] = []
```
Comment (heheda12345) [inline vllm/v1/request.py]: Where do you calculate the block hash for the prompt?

File: vllm/v1/core/sched/scheduler.py (hunk 2 above)
Comment (orozery) [reply vllm/v1/core/sched/scheduler.py]: And import `generate_block_hash_extra_keys`, `hash_block_tokens` from inside `append_output_token_ids`?
How would that function get access to `block_size` and `caching_hash_fn`?

File: vllm/v1/request.py (hunk 3 above)
Comment (orozery) [reply vllm/v1/request.py]: In here:
https://github.com/vllm-project/vllm/blob/ac256d328fb39f9e74c51a06d9a0e6e5944d6a10/vllm/v1/engine/core.py#L242

File: vllm/utils/__init__.py [hunk 4]
```diff
@@ -3236,6 +3236,25 @@ def sha256_cbor_64bit(input) -> int:
     return full_hash & ((1 << 64) - 1)
 
 
+@lru_cache
```
Comment (heheda12345) [inline vllm/utils/__init__.py]: I think we don't need a cache. This function should only be called once during initialization.

File: vllm/v1/core/sched/scheduler.py (hunk 2 above)
Comment (heheda12345) [reply vllm/v1/core/sched/scheduler.py]: Is it ok to add the hash function as an attribute of a request? Then we can update the hash during constructing the request and append_output_token_ids

**Comment by orozery** [comment]
@heheda12345 @njhill Moved all hashing to be triggered by `Request` directly.

**Comment by njhill** [comment]
@orozery some test failures are due to the changes. It would probably be best to have `medium` default to `None`.

File: vllm/v1/core/kv_cache_utils.py [hunk 5]
```diff
@@ -547,41 +547,50 @@ def hash_block_tokens(
...
-    return ret
+
+    def request_block_hasher(request: Request) -> list[BlockHash]:
+        start_token_idx = len(request.block_hashes) * block_size
+        num_tokens = request.num_tokens
+
+        curr_mm_idx = 0
+        if start_token_idx > 0:
+            curr_mm_idx = -1
+            assert start_token_idx + 2 * block_size > num_tokens
```
Comment (heheda12345) [inline vllm/v1/core/kv_cache_utils.py]: why do you need this constraint? what if `block_size=1` and spec decode accepts 4 tokens?

File: vllm/v1/core/kv_cache_utils.py [hunk 6]
```diff
@@ -547,41 +547,50 @@ def hash_block_tokens(
...
-        parent_block_hash_value = block_hash.hash_value
-    return ret
+
+    def request_block_hasher(request: Request) -> list[BlockHash]:
+        start_token_idx = len(request.block_hashes) * block_size
+        num_tokens = request.num_tokens
+
+        curr_mm_idx = 0
+        if start_token_idx > 0:
+            curr_mm_idx = -1
```
Comment (heheda12345) [inline vllm/v1/core/kv_cache_utils.py:577]: Can you add some comments. I've copied some comments that I think is worth mention here. Please adjust them for the current context. 
1.
```
                # Generate extra keys for multi-modal inputs. Note that since
                # we reach to this branch only when the block is completed with
                # generated tokens, we only need to consider the last mm input.
````
2. `    # Support start_mm_idx == -1 to indicate the last mm input.`

File: vllm/v1/core/kv_cache_utils.py [hunk 7]
```diff
@@ -547,41 +547,50 @@ def hash_block_tokens(
...
+            assert start_token_idx + 2 * block_size > num_tokens
+
+        prev_block_hash_value = request.block_hashes[-1].hash_value \
+            if request.block_hashes else None
+        new_block_hashes: list[BlockHash] = []
+        while True:
+            end_token_idx = start_token_idx + block_size
```
Comment (heheda12345) [inline vllm/v1/core/kv_cache_utils.py:583]:
```suggestion
            end_token_idx = start_token_idx + block_size
            # Only full blocks will be hashed
```

File: vllm/v1/core/kv_cache_utils.py [hunk 8]
```diff
@@ -547,41 +547,50 @@ def hash_block_tokens(
...
+            if end_token_idx > num_tokens:
+                break
+            extra_keys, curr_mm_idx = generate_block_hash_extra_keys(
```
Comment (heheda12345) [inline vllm/v1/core/kv_cache_utils.py:589]:
```suggestion
            # MM and LoRA requests need extra keys for block-hash computation.
            extra_keys, curr_mm_idx = generate_block_hash_extra_keys(
```

File: vllm/v1/request.py [hunk 9]
```diff
@@ -108,8 +111,17 @@ def __init__(
         # indicates that the output is corrupted
         self.num_nans_in_logits = 0
 
+        self.block_hashes: list[BlockHash] = []
+        self.block_hasher: Optional[Callable[[Request],
```
Comment (heheda12345) [inline vllm/v1/request.py]: Personally prefer `self.block_hasher= partial(block_hasher, request=self)` so that we can call it with `self.block_hasher()` without passing `self`.

File: vllm/v1/request.py [hunk 10]
```diff
@@ -145,6 +158,9 @@ def append_output_token_ids(
             self._output_token_ids.extend(token_ids)
             self._all_token_ids.extend(token_ids)
 
+        if self.block_hasher is not None:
+            self.block_hashes += self.block_hasher(self)
```
Comment (heheda12345) [inline vllm/v1/request.py]: What about calling it `self.get_hash_new_full_blocks` so that people will know it only applies on new blocks instead of re-compute the hash of the full request?

And can you either
1. use *.extend() just like the above lines, or
2. put `+=` logic into the function (then maybe change the function name to `self.hash_new_full_blocks()`).
I'm fine with both choices and it's up to you.

**Review by heheda12345 — COMMENTED** [review]
Sorry for my late response. Can you add more comments in `hash_request_tokens` and `cache_full_blocks`. I think there are many tricks that needs to be explained here

**Comment by orozery** [comment]
> Can you add more comments in `hash_request_tokens`

You mean in `get_request_block_hasher`

**Comment by heheda12345** [comment]
I want to say move some comments in `hash_request_tokens` and `cache_full_blocks` to your new code.

**Comment by orozery** [comment]
@heheda12345 I made another set of changes.
Thanks for taking the time giving out really helpful comments!

**Review by heheda12345 — APPROVED** [review]
LGTM! Thanks for your great work and patience.

**Comment by njhill** [comment]
@orozery could you rebase on latest main? Should hopefully help with at least some of the CI failures.

**Comment by orozery** [comment]
> @orozery could you rebase on latest main? Should hopefully help with at least some of the CI failures.

done

**Comment by heheda12345** [comment]
Can you rebase with main again? It should fix the test_eagle_correctness failure.
